test/e2e/e2e.go

```diff
@@ -40,3 +45,83 @@ func SetupWithNamespace(t *testing.T, namespace string) *test.Clients {
 	}
 	return clients
 }
+
+// knativeServingVerify verifies if the KnativeServing can reach the READY status.
+func knativeServingVerify(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	if _, err := resources.WaitKnativeServingReady(t, clients.KnativeServingAlphaClient, names.KnativeServing,
+		resources.IsKnativeServingReady); err != nil {
+		t.Fatalf("KnativeService %q failed to get to the READY status: %v", names.KnativeServing, err)
+	}
+
+}
+
+// deploymentRecreation verify whether all the deployments for knative serving are able to recreate, when they are deleted.
+func deploymentRecreation(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	dpListSave, err := clients.KubeClient.Kube.AppsV1().Deployments(names.Namespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get any deployment under the namespace %q: %v",
+			test.ServingOperatorNamespace, err)
+	}
+	if len(dpListSave.Items) == 0 {
+		t.Fatalf("No deployment under the namespace %q was found",
+			test.ServingOperatorNamespace)
+	}
+	// Delete the deployments one by one to see if they will be recreated.
+	for _, deployment := range dpListSave.Items {
+		if err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Delete(deployment.Name,
+			&metav1.DeleteOptions{}); err != nil {
+			t.Fatalf("Failed to delete deployment %s/%s: %v", deployment.Namespace, deployment.Name, err)
+		}
+
+		waitErr := wait.PollImmediate(resources.Interval, resources.Timeout, func() (bool, error) {
+			dep, err := clients.KubeClient.Kube.AppsV1().Deployments(deployment.Namespace).Get(deployment.Name, metav1.GetOptions{})
+			if err != nil && apierrs.IsNotFound(err) {
+				// If the deployment is not found, we continue to wait for the availability.
+				return false, nil
+			}
+			return resources.IsDeploymentAvailable(dep)
+		})
+
+		if waitErr != nil {
+			t.Fatalf("The deployment %s/%s failed to reach the desired state: %v", deployment.Namespace, deployment.Name, err)
+		}
+
+		if _, err := resources.WaitForKnativeServingState(clients.KnativeServingAlphaClient, test.ServingOperatorName,
+			resources.IsKnativeServingReady); err != nil {
+			t.Fatalf("KnativeService %q failed to reach the desired state: %v", test.ServingOperatorName, err)
+		}
+		t.Logf("The deployment %s/%s reached the desired state.", deployment.Namespace, deployment.Name)
+	}
+}
+
+// knativeServingDeletion deletes tha KnativeServing to see if all the deployments will be removed.
+func knativeServingDeletion(t *testing.T, clients *test.Clients, names test.ResourceNames) {
+	if err := clients.KnativeServingAlphaClient.Delete(names.KnativeServing, &metav1.DeleteOptions{}); err != nil {
+		t.Fatalf("KnativeService %q failed to delete: %v", names.KnativeServing, err)
+	}
+	if _, err := resources.WaitForKnativeServingState(clients.KnativeServingAlphaClient, names.KnativeServing,
```

Comment:
Why do we need to wait for this state? Is this delayed because of finalizers? I'd think the above would block until it's really really deleted so checking its error might suffice?

Reply:
OK. If the resource is removed before clients.KnativeServingAlphaClient.Delete returns, then we do not need to verify the status of WaitForKnativeServingState.